straight.el

```
@@ -4260,11 +4260,15 @@ The name of the cache file is stored in
   (straight--log 'modification-detection "Saving build cache")
   (unless straight-safe-mode
     (with-temp-buffer
-      ;; Prevent mangling of the form being printed in the case that
-      ;; this function was called by an `eval-expression' invocation
-      ;; of `straight-use-package'.
-      (let ((print-level nil)
-            (print-length nil))
+      (let (;; Prevent mangling of the form being printed in the case that
+            ;; this function was called by an `eval-expression' invocation
+            ;; of `straight-use-package'.
+            (print-level nil)
+            (print-length nil)
+            ;; Print symbols with attached positions, which can occur
+            ;; during byte compilation, as bare symbols so they can be
+            ;; read when loading the cache.
+            (print-symbols-bare t))
```

```suggestion
            (print-symbols-bare t))
          (ignore print-symbols-bare)
```

It looks like that variable was only introduce in Emacs 29.
That's why the CI is failing with a byte-compiler warning:

https://github.com/radian-software/straight.el/actions/runs/7953631018/job/21709882276?pr=1149#step:3:2400

---

That (at least partly) explains why I only noticed this problem recently!

---

> You could imagine why something like this might be plausible when the package needs to be installed at macro expansion time for whatever reason.

I'm not sure I follow. When/why would one want to install packages during macro expansion?
@raxod502 any thoughts on this use case?

> Interestingly, I couldn't get this to reproduce using eval-and-compile. This is because eval ignores positions when symbols-with-pos-enabled is non-nil (as it is during byte-compilation).

Why not have the macro expansion make use of `eval-and-compile` in that case?:

```emacs-lisp
(defmacro test-macro (&rest args)
  `(eval-and-compile ...))
```

---

> I'm not sure I follow. When/why would one want to install packages during macro expansion? @raxod502 any thoughts on this use case?

Well, for example [use-package](https://github.com/jwiegley/use-package/blob/ff8c57268a3b655098bd9cfd05e9d4a689533c3a/use-package-ensure.el#L192) does this. The reason is that the package may define macros which are used subsequently in the file. To ensure the later macros can be expanded, the package must be installed at compile time.

> Why not have the macro expansion make use of `eval-and-compile` in that case?:

I think it should be possible to use `eval-and-compile` or possibly `eval-when-compile` to achieve the desired effect, but since the fix is simple and self contained, I don't see a reason not to accommodate users who are not using those macros (or use packages that do not). Especially so given that the unpatched behavior 1. isn't obviously correct and 2. leads to a silent failure that is difficult to diagnose.

---

To be clear, I'm not against the solution. I'm just trying to understand the problem fully before we apply a fix.
That byte compilation warning still needs to be addressed before we can merge, too.

---

Got it! Let me take a stab at fixing the warning and see if it's to your liking.

---

Installing packages during byte-compilation doesn't seem like a great idea to me, but it also doesn't seem so wrong that I would say we should not allow it. Of course we should fix the bug so I agree with the submitted patch.